### DISCUSSION TITLE
Announcing typebox-validators

### DISCUSSION THREAD

**Discussion by jtlapp** [sinclairzx81/typebox#471] [Show and tell]
I love TypeBox, but TypeBox limits itself to standard behavior, so I created a package of validators that embellish TypeBox and provide useful convenience methods. The result is the package [`typebox-validators`](https://github.com/jtlapp/typebox-validators), providing validators with lazy compilation, custom error messages, safe error handling, discriminated and heterogeneous unions.

It provides these validator classes:

- StandardValidator
- CompilingStandardValidator
- DiscriminatedUnionValidator
- CompilingDiscriminatedUnionValidator
- HeterogeneousUnionValidator
- CompilingHeterogeneousUnionValidator

The "compiling" validators compile TypeBox schemas on their first usage and cache the compilations for subsequent use.

Here is a sample:

```ts
import { Type } from '@sinclaim/typebox';
import { StandardValidator } from 'typebox-validators/standard';

const schema = Type.Object({
  handle: Type.String(
    {
      minLength: 5,
      maxLength: 10,
      pattern: '^[a-zA-Z]+$',
      errorMessage: 'must be a string of 5 to 10 letters',
    }
  ),
  count: Type.Integer({ minimum: 0 }),
});

const validator = new StandardValidator(schema);
const value = { handle: '1234', count: -1 };

// returns false indicating an error:
validator.test(value);

// after calling `test`, returns an iterable providing the two errors:
validator.testReturningErrors();

// after calling `test`, returns the first error:
validator.testReturningFirstError();

// throws with error 'must be a string of 5 to 10 letters':
validator.assert(value);

// throws with error 'must be a string of 5 to 10 letters' and TypeBox's
//  default error message for an integer being less than the minimum:
validator.validate(value);

// returns an iterable providing the two errors:
validator.errors(value);

// returns the first error:
validator.firstError(value);
```

**Comment by sinclairzx81** [comment]
@jtlapp Hey! That's awesome!! (especially the support for custom `errorMessage`), great work! I :)

TypeBox is tracking community packages under the [Ecosystem](https://github.com/sinclairzx81/typebox#ecosystem) section of the readme. Would you like to include `typebox-validators` in the list? I think your project would be a great addition (as quite a few people ask about custom errors as well as discriminated / tagged unions). Being able to refer them to your project would be good when the topic comes up.

Again, great work!
S

**Reply by jtlapp** [reply]
Oh yeah, please do. Thanks so much!

**Reply by jtlapp** [reply]
Hey @sinclairzx81, when I return a custom error message, should I leave `type` as that of the first detected error, or should I change it to `ValueErrorType.Custom`? The error message replaces all errors that might occur, not just the first error, collapsing multiple errors into a single message.

I'm getting ready to submit a PR for React Hook Form's [typeboxResolver](https://github.com/react-hook-form/resolvers#TypeBox) providing custom error messages, and I'll also need to set `type` appropriately there.
